greenplumpython/func.py

```diff
@@ -170,292 +158,332 @@ def serialize(self) -> str:
+_global_scope: Dict[str, _AbstractFunction] = {}
 
-    return make_function_call
 
+class NormalFunction(_AbstractFunction):
+    def __init__(
+        self,
+        wrapped_func: Optional[Callable[..., Any]] = None,
+        name: Optional[str] = None,
+        schema: Optional[str] = None,
+        returning_composite: Optional[bool] = False,
+        language_handler: str = "plpython3u",
+    ) -> None:
+        super().__init__(wrapped_func, name, schema)
+        self._created_in_dbs: Optional[Set[Database]] = set() if wrapped_func is not None else None
+        self._wrapped_func = wrapped_func
+        self._language_handler = language_handler
+        self._returning_composite = returning_composite
 
-def aggregate(name: str) -> Callable[..., FunctionExpr]:
+    def unwrap(self) -> Callable[..., Any]:
+        """
+        Get the wrapped Python function in the database function.
+        """
+        assert self._wrapped_func is not None, "No Python function is wrapped inside."
+        return self._wrapped_func
+
+    def _create_in_db(self, db: Database) -> None:
+        assert self._created_in_dbs is not None
+        assert self._wrapped_func is not None
+        if db not in self._created_in_dbs:
+            func_sig = inspect.signature(self._wrapped_func)
+            func_args = ",".join(
+                [
+                    f"{param.name} {to_pg_type(param.annotation, db)}"
+                    for param in func_sig.parameters.values()
+                ]
+            )
+            # -- Loading Python code of Function
+            # FIXME: include things in func.__closure__
+            func_lines = textwrap.dedent(inspect.getsource(self._wrapped_func)).split("\n")
+            func_body = "\n".join([line for line in func_lines if re.match(r"^\s", line)])
+            return_type = to_pg_type(func_sig.return_annotation, db, for_return=True)
+            self._returning_composite = func_sig.return_annotation not in primitive_type_map
+            assert (
+                db.execute(
+                    (
+                        f"CREATE FUNCTION {self._qualified_name} ({func_args}) "
+                        f"RETURNS {return_type} "
+                        f"AS $$\n"
+                        f"{textwrap.dedent(func_body)} $$"
+                        f"LANGUAGE {self._language_handler};"
+                    ),
+                    has_results=False,
+                )
+                is None
+            )
+            self._created_in_dbs.add(db)
+
+    def __call__(self, *args: Any, db: Optional[Database] = None) -> FunctionExpr:
+        if self._wrapped_func is not None:
+            _db = self._infer_db(args, db)
+            assert _db is not None, "Database is required to create function"
+            self._create_in_db(_db)
+        assert (
+            self._returning_composite is not None
+        ), f'Unknown if the return type of function "{self.qualified_name}" is composite.'
```

Comment:
```suggestion
        ), f'The return type of function "{self.qualified_name}" is expected to be composite.'
```

Maybe a better error message?